pkg/handler/v2/cluster/serviceaccount.go

```diff
@@ -0,0 +1,324 @@
+}
+
+func DeleteClusterSAKubeconigEndpoint(userInfoGetter provider.UserInfoGetter, projectProvider provider.ProjectProvider, privilegedProjectProvider provider.PrivilegedProjectProvider) endpoint.Endpoint {
+	return func(ctx context.Context, request interface{}) (interface{}, error) {
+		req := request.(ClusterSAReq)
+		clusterProvider := ctx.Value(middleware.ClusterProviderContextKey).(provider.ClusterProvider)
+
+		cluster, err := handlercommon.GetCluster(ctx, projectProvider, privilegedProjectProvider, userInfoGetter, req.ProjectID, req.ClusterID, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		client, err := common.GetClusterClient(ctx, userInfoGetter, clusterProvider, cluster, req.ProjectID)
+		if err != nil {
+			return nil, common.KubernetesErrorToHTTPError(err)
+		}
+
+		serviceAccount := &corev1.ServiceAccount{}
+		if err := client.Get(ctx, types.NamespacedName{Namespace: req.Namespace, Name: req.ServiceAccountID}, serviceAccount); err != nil {
+			return nil, common.KubernetesErrorToHTTPError(err)
+		}
+		if serviceAccount.GetLabels()[ServiceAccountComponentKey] != ServiceAccountComponentValue {
+			return nil, utilerrors.New(http.StatusForbidden, fmt.Sprintf("can not delete service account which is not labeled %s=%s", ServiceAccountComponentKey, ServiceAccountComponentValue))
+		}
+
+		if err := handlercommon.UnbindServiceAccountFromRoles(ctx, client, serviceAccount); err != nil {
+			return nil, err
+		}
+
+		if err := handlercommon.UnbindServiceAccountFromClusterRoles(ctx, client, serviceAccount); err != nil {
+			return nil, err
+		}
+
+		if err := client.Delete(ctx, serviceAccount); err != nil {
+			return nil, common.KubernetesErrorToHTTPError(err)
+		}
+		return nil, nil
+	}
+}
+
+func convertInternalClusterSAToExternal(sa *corev1.ServiceAccount) *apiv2.ClusterServiceAccount {
+	return &apiv2.ClusterServiceAccount{
+		ObjectMeta: apiv1.ObjectMeta{
+			ID:                sa.Name,
+			Name:              sa.Name,
+			CreationTimestamp: apiv1.NewTime(sa.CreationTimestamp.Time),
+		},
+		Namespace: sa.Namespace,
+	}
+}
+
+// listClusterSAReq defines HTTP request for ListClusterSAEndpoint
+// swagger:parameters listClusterServiceAccount
+type listClusterSAReq struct {
+	common.ProjectReq
+	// in: path
+	// required: true
+	ClusterID string `json:"cluster_id"`
+}
+
+func (req listClusterSAReq) GetSeedCluster() apiv1.SeedCluster {
+	return apiv1.SeedCluster{
+		ClusterID: req.ClusterID,
+	}
+}
+
+func DecodeListClusterServiceAccount(c context.Context, r *http.Request) (interface{}, error) {
+	var req listClusterSAReq
+
+	projectReq, err := common.DecodeProjectRequest(c, r)
+	if err != nil {
+		return nil, err
+	}
+	req.ProjectReq = projectReq.(common.ProjectReq)
+
+	clusterID, err := common.DecodeClusterID(c, r)
+	if err != nil {
+		return nil, err
+	}
+	req.ClusterID = clusterID
+
+	return req, nil
+}
+
+// Validate validates listClusterSAReq request.
+func (req listClusterSAReq) Validate() error {
+	if len(req.ProjectID) == 0 {
```

Comment:
Is there any reason why `ClusterID` is not checked same way as `ProjectID`?

Reply:
opposite to `ProjectID` validation is already done in [DecodeClusterID](https://github.com/kubermatic/kubermatic/blob/84b83afc7770e15f18a44f0cc9e57d2bb9250656/pkg/handler/v1/common/request.go#L127-L129) function.